File: api/generate-pdf.py
from http.server import BaseHTTPRequestHandler
import json
from io import BytesIO
from reportlab.lib.pagesizes import letter
from reportlab.platypus import SimpleDocTemplate, Paragraph, Spacer, Table, TableStyle, Image
from reportlab.lib.styles import getSampleStyleSheet, ParagraphStyle
from reportlab.lib import colors

class handler(BaseHTTPRequestHandler):
    def do_POST(self):
        content_length = int(self.headers.get('Content-Length', 0))
        post_data = self.rfile.read(content_length)
        data = json.loads(post_data.decode('utf-8'))
        
        pdf_type = data.get('type', 'attendance')
        title = data.get('title', 'Event')
        coordinator = data.get('coordinator', 'N/A')
        date = data.get('date', 'N/A')
        time = data.get('time', 'N/A')
        venue = data.get('venue', 'N/A')
        students = data.get('students', [])
        
        # Create a file-like buffer to receive PDF data
        buffer = BytesIO()
        
        # Setup document (portrait, 0.5 inch margins)
        doc = SimpleDocTemplate(
            buffer,
            pagesize=letter,
            rightMargin=36,
            leftMargin=36,
            topMargin=36,
            bottomMargin=36
        )
        
        story = []
        styles = getSampleStyleSheet()
        
        # Define styles
        title_style = ParagraphStyle(
            'TitleStyle',
            parent=styles['Heading2'],
            fontName='Helvetica-Bold',
            fontSize=13,
            leading=16,
            textColor=colors.HexColor('#000000'),
            alignment=1 # Center
        )
        subtitle_style = ParagraphStyle(
            'SubtitleStyle',
            parent=styles['Normal'],
            fontName='Helvetica-Bold',
            fontSize=10,
            leading=13,
            textColor=colors.HexColor('#111111'),
            alignment=1 # Center
        )
        
        body_style = ParagraphStyle(
            'BodyStyle',
            parent=styles['Normal'],
            fontName='Helvetica',
            fontSize=8.5,
            leading=11
        )
        
        header_cell_style = ParagraphStyle(
            'HeaderCellStyle',
            parent=styles['Normal'],
            fontName='Helvetica-Bold',
            fontSize=8.5,
            leading=11,
            textColor=colors.white,
            alignment=1 # Center
        )
        
        # 1. Header (College Name & Logo)
        import os
        logo_path = os.path.join(os.path.dirname(__file__), '..', 'college_logo.png')
        event_logo_path = os.path.join(os.path.dirname(__file__), '..', 'event_logo.png')
        
        report_titles = {
            'attendance': 'REGISTRANTS DIRECTORY & ATTENDANCE SHEET',
            'registrations': 'REGISTRATIONS DIRECTORY',
            'marksheet': f"OFFICIAL JUDGING MARKSHEET - {data.get('round', '').upper()}" if data.get('round') else 'OFFICIAL JUDGING MARKSHEET'
        }
        report_title = report_titles.get(pdf_type, 'EVENT REPORT')
        
        header_table = None
        has_logo = os.path.exists(logo_path)
        has_event_logo = os.path.exists(event_logo_path)
        
        try:
            if has_logo and has_event_logo:
                logo_img = Image(logo_path, width=40, height=40)
                event_logo_img = Image(event_logo_path, width=40, height=40)
                
                header_text = [
                    Paragraph("DR. B.B HEGDE FIRST GRADE COLLEGE, KUNDAPURA", ParagraphStyle('T1', parent=title_style, fontSize=11, leading=13)),
                    Spacer(1, 2),
                    Paragraph("TECH MANTHAN 6.0", ParagraphStyle('S1', parent=subtitle_style, fontSize=9, leading=11)),
                    Spacer(1, 1),
                    Paragraph(report_title, ParagraphStyle('R1', parent=subtitle_style, fontSize=8, fontName='Helvetica-Oblique', textColor=colors.HexColor('#555555')))
                ]
                header_table = Table([[logo_img, header_text, event_logo_img]], colWidths=[45, 360, 45], hAlign='CENTER')
                header_table.setStyle(TableStyle([
                    ('VALIGN', (0,0), (-1,-1), 'MIDDLE'),
                    ('ALIGN', (0,0), (-1,-1), 'CENTER'),
                    ('LEFTPADDING', (0,0), (-1,-1), 0),
                    ('RIGHTPADDING', (0,0), (-1,-1), 0),
                    ('BOTTOMPADDING', (0,0), (-1,-1), 0),
                    ('TOPPADDING', (0,0), (-1,-1), 0),
                ]))
            elif has_logo:
                logo_img = Image(logo_path, width=40, height=40)
                header_text = [
                    Paragraph("DR. B.B HEGDE FIRST GRADE COLLEGE, KUNDAPURA", ParagraphStyle('T2', parent=title_style, fontSize=11, leading=13, alignment=0)),
                    Spacer(1, 2),
                    Paragraph("TECH MANTHAN 6.0", ParagraphStyle('S2', parent=subtitle_style, fontSize=9, leading=11, alignment=0)),
                    Spacer(1, 1),
                    Paragraph(report_title, ParagraphStyle('R2', parent=subtitle_style, fontSize=8, fontName='Helvetica-Oblique', textColor=colors.HexColor('#555555'), alignment=0))
                ]
                header_table = Table([[logo_img, header_text]], colWidths=[45, 405], hAlign='CENTER')
                header_table.setStyle(TableStyle([
                    ('VALIGN', (0,0), (-1,-1), 'MIDDLE'),
                    ('ALIGN', (0,0), (-1,-1), 'LEFT'),
                    ('LEFTPADDING', (0,0), (-1,-1), 0),
                    ('RIGHTPADDING', (0,0), (-1,-1), 0),
                    ('BOTTOMPADDING', (0,0), (-1,-1), 0),
                    ('TOPPADDING', (0,0), (-1,-1), 0),
                ]))
            else:
                header_text = [
                    Paragraph("DR. B.B HEGDE FIRST GRADE COLLEGE, KUNDAPURA", ParagraphStyle('T3', parent=title_style, fontSize=11, leading=13)),
                    Spacer(1, 2),
                    Paragraph("TECH MANTHAN 6.0", ParagraphStyle('S3', parent=subtitle_style, fontSize=9, leading=11)),
                    Spacer(1, 1),
                    Paragraph(report_title, ParagraphStyle('R3', parent=subtitle_style, fontSize=8, fontName='Helvetica-Oblique', textColor=colors.HexColor('#555555')))
                ]
                header_table = Table([[header_text]], colWidths=[450], hAlign='CENTER')
                header_table.setStyle(TableStyle([
                    ('VALIGN', (0,0), (-1,-1), 'MIDDLE'),
                    ('ALIGN', (0,0), (-1,-1), 'CENTER'),
                ]))
            story.append(header_table)
        except Exception as e:
            story.append(Paragraph("DR. B.B HEGDE FIRST GRADE COLLEGE, KUNDAPURA", title_style))
            story.append(Spacer(1, 2))
            story.append(Paragraph("TECH MANTHAN 6.0", subtitle_style))
            story.append(Spacer(1, 1))
            story.append(Paragraph(report_title, subtitle_style))

        story.append(Spacer(1, 10))
        
        # 2. Event Metadata Box
        meta_data = [
            [
                Paragraph(f"<b>Event Name:</b> {title}", body_style),
                Paragraph(f"<b>Coordinator:</b> {coordinator}", body_style)
            ],
            [
                Paragraph(f"<b>Date & Time:</b> {date} | {time}", body_style),
                Paragraph(f"<b>Venue:</b> {venue}", body_style)
            ]
        ]
        meta_table = Table(meta_data, colWidths=[270, 270])
        meta_table.setStyle(TableStyle([
            ('BACKGROUND', (0,0), (-1,-1), colors.HexColor('#f8fafc')),
            ('BOX', (0,0), (-1,-1), 0.5, colors.HexColor('#cbd5e1')),
            ('INNERGRID', (0,0), (-1,-1), 0.5, colors.HexColor('#e2e8f0')),
            ('PADDING', (0,0), (-1,-1), 6),
            ('VALIGN', (0,0), (-1,-1), 'MIDDLE'),
        ]))
        story.append(meta_table)
        story.append(Spacer(1, 12))
        
        # 3. Students Table
        if data.get('headers') and data.get('rows'):
            # Dynamic Excel-style PDF table
            raw_headers = data.get('headers')
            raw_rows = data.get('rows')
            
            table_data = [[Paragraph(str(h), header_cell_style) for h in raw_headers]]
            for row in raw_rows:
                table_data.append([Paragraph(str(cell), body_style) for cell in row])
                
            num_cols = len(raw_headers)
            available_width = 540
            col_widths = [available_width / num_cols] * num_cols
        elif pdf_type in ['attendance', 'registrations']:
            # Attendance columns: Sl No, Reg No, Student Name, Class, Email, Attendance, Signature
            table_data = [[
                Paragraph("Sl No", header_cell_style),
                Paragraph("Reg No", header_cell_style),
                Paragraph("Student Name", header_cell_style),
                Paragraph("Class", header_cell_style),
                Paragraph("Email Address", header_cell_style),
                Paragraph("Attendance", header_cell_style),
                Paragraph("Signature / Sign", header_cell_style)
            ]]
            
            for idx, st in enumerate(students):
                status = "Present" if st.get('checkedIn', False) else "Absent"
                status_color = "#16a34a" if st.get('checkedIn', False) else "#dc2626"
                
                table_data.append([
                    Paragraph(str(idx + 1), body_style),
                    Paragraph(f"<b>{st.get('regNo', '')}</b>", body_style),
                    Paragraph(st.get('name', 'N/A'), body_style),
                    Paragraph(st.get('class', 'N/A'), body_style),
                    Paragraph(st.get('email', 'N/A'), body_style),
                    Paragraph(f"<font color='{status_color}'><b>{status}</b></font>", body_style),
                    Paragraph("", body_style) # Empty for signature
                ])
                
            col_widths = [30, 70, 110, 70, 120, 60, 80]
        else:
            # Marksheet columns: Sl No, Reg No, Student Name, Evaluation Breakdown, Final Score
            table_data = [[
                Paragraph("Sl No", header_cell_style),
                Paragraph("Reg No", header_cell_style),
                Paragraph("Student Name", header_cell_style),
                Paragraph("Judge Evaluation Breakdowns", header_cell_style),
                Paragraph("Final Average Score", header_cell_style)
            ]]
            
            for idx, st in enumerate(students):
                table_data.append([
                    Paragraph(str(idx + 1), body_style),
                    Paragraph(f"<b>{st.get('regNo', '')}</b>", body_style),
                    Paragraph(st.get('name', 'N/A'), body_style),
                    Paragraph(st.get('breakdownText', 'N/A'), body_style),
                    Paragraph(f"<b>{st.get('avgTotal', 0)} pts</b>", body_style)
                ])
                
            col_widths = [30, 80, 130, 210, 90]
            
        student_table = Table(table_data, colWidths=col_widths)
        student_table.setStyle(TableStyle([
            ('BACKGROUND', (0,0), (-1,0), colors.HexColor('#5b21b6')), # Purple header
            ('ALIGN', (0,0), (-1,-1), 'LEFT'),
            ('VALIGN', (0,0), (-1,-1), 'MIDDLE'),
            ('PADDING', (0,0), (-1,-1), 8),
            ('GRID', (0,0), (-1,-1), 0.5, colors.HexColor('#dddddd')),
            ('ROWBACKGROUNDS', (0,1), (-1,-1), [colors.white, colors.HexColor('#fafafa')]),
        ]))
        story.append(student_table)
        
        # 4. Footer signature
        story.append(Spacer(1, 30))
        footer_style = ParagraphStyle(
            'FooterStyle',
            parent=styles['Normal'],
            fontName='Helvetica-Oblique',
            fontSize=9,
            alignment=2 # Right
        )
        story.append(Paragraph("Coordinator Signature: _______________________", footer_style))
        
        # Build PDF
        doc.build(story)
        
        pdf_data = buffer.getvalue()
        buffer.close()
        
        self.send_response(200)
        self.send_header('Content-type', 'application/pdf')
        self.send_header('Content-Disposition', f'attachment; filename="{pdf_type}_{title.lower().replace(" ", "_")}.pdf"')
        self.send_header('Content-Length', str(len(pdf_data)))
        self.end_headers()
        self.wfile.write(pdf_data)

File: api/generate-pdf.js
const PDFDocument = require('pdfkit');
const path = require('path');
const fs = require('fs');

module.exports = async function handleGeneratePDF(req, res) {
  try {
    const payload = req.body || {};
    const { type, title, coordinator, date, time, venue, round, judge, students, headers, rows } = payload;

    // Build standard headers and rows if type is attendance or registrations
    let tableHeaders = headers || [];
    let tableRows = rows || [];

    if ((type === 'attendance' || type === 'registrations') && Array.isArray(students)) {
      tableHeaders = ["Sl No", "Reg No", "Student Name", "Class", "Email Address", "Status"];
      tableRows = students.map((st, idx) => [
        idx + 1,
        st.regNo || "N/A",
        st.name || "N/A",
        st.class || "N/A",
        st.email || "N/A",
        st.checkedIn ? "PRESENT" : "ABSENT"
      ]);
    }

    // Determine layout: landscape if more than 6 columns
    const isLandscape = tableHeaders.length > 6;
    const doc = new PDFDocument({
      size: 'A4',
      layout: isLandscape ? 'landscape' : 'portrait',
      margin: 30,
      bufferPages: true
    });

    const buffers = [];
    doc.on('data', chunk => buffers.push(chunk));
    doc.on('end', () => {
      const pdfBuffer = Buffer.concat(buffers);
      res.writeHead(200, {
        'Content-Type': 'application/pdf',
        'Content-Length': pdfBuffer.length,
        'Content-Disposition': 'inline; filename="report.pdf"'
      });
      res.end(pdfBuffer);
    });

    const pageWidth = isLandscape ? 841.89 : 595.28;
    const pageHeight = isLandscape ? 595.28 : 841.89;
    const margin = 30;
    const contentWidth = pageWidth - (margin * 2);

    // Header Background Accent Bar
    doc.rect(margin, margin, contentWidth, 54).fill('#0f172a');

    // College & Event Header Text
    doc.fillColor('#00f3ff')
      .font('Helvetica-Bold')
      .fontSize(14)
      .text('TECH MANTHAN 6.0', margin + 15, margin + 8, { width: contentWidth - 30, align: 'left' });

    doc.fillColor('#ffffff')
      .font('Helvetica-Bold')
      .fontSize(10)
      .text('Dr. B. B. HEGDE FIRST GRADE COLLEGE, KUNDAPURA', margin + 15, margin + 26, { width: contentWidth - 30, align: 'left' });

    let subTitleText = (title || 'EVENT REPORT').toUpperCase();
    if (type === 'attendance') subTitleText = `ATTENDANCE REPORT: ${subTitleText}`;
    else if (type === 'registrations') subTitleText = `REGISTRATIONS DIRECTORY: ${subTitleText}`;
    else if (type === 'marksheet') subTitleText = `MARKSHEET REPORT: ${subTitleText} ${round ? `(${round.toUpperCase()})` : ''}`;

    doc.fillColor('#cbd5e1')
      .font('Helvetica')
      .fontSize(8)
      .text(subTitleText, margin + 15, margin + 38, { width: contentWidth - 30, align: 'left' });

    let startY = margin + 64;

    // Metadata Summary Box (2 Rows for clean spacing)
    const metaBoxHeight = 36;
    doc.rect(margin, startY, contentWidth, metaBoxHeight).fill('#f1f5f9').stroke('#cbd5e1');

    // Row 1: Date, Time, Venue, Judge
    doc.fillColor('#334155').font('Helvetica-Bold').fontSize(8.5);
    const metaPartsRow1 = [];
    if (date) metaPartsRow1.push(`Date: ${date}`);
    if (time) metaPartsRow1.push(`Time: ${time}`);
    if (venue) metaPartsRow1.push(`Venue: ${venue}`);
    if (judge) metaPartsRow1.push(`Judge: ${judge}`);

    doc.text(metaPartsRow1.join('   |   '), margin + 10, startY + 6, { width: contentWidth - 20, align: 'left' });

    // Row 2: Event Coordinator
    doc.fillColor('#0f172a').font('Helvetica-Bold').fontSize(8.5);
    doc.text(`Event Coordinator: ${coordinator || 'Unassigned'}`, margin + 10, startY + 20, { width: contentWidth - 20, align: 'left' });

    startY += metaBoxHeight + 12;

    // Render Table
    if (tableHeaders.length > 0) {
      // Calculate Column Widths dynamically
      const colCount = tableHeaders.length;
      let colWidths = [];

      if (type === 'attendance' || type === 'registrations') {
        colWidths = [
          contentWidth * 0.08, // Sl No
          contentWidth * 0.18, // Reg No
          contentWidth * 0.26, // Name
          contentWidth * 0.16, // Class
          contentWidth * 0.20, // Email
          contentWidth * 0.12  // Status
        ];
      } else {
        colWidths = tableHeaders.map((h, i) => {
          const text = String(h).toLowerCase();
          if (text.includes('name')) return (contentWidth / colCount) * 1.4;
          if (text.includes('sl') || text.includes('no')) return (contentWidth / colCount) * 0.6;
          return contentWidth / colCount;
        });
        const currentSum = colWidths.reduce((a, b) => a + b, 0);
        colWidths = colWidths.map(w => (w / currentSum) * contentWidth);
      }

      // Draw Table Header
      const headerHeight = 22;
      doc.rect(margin, startY, contentWidth, headerHeight).fill('#1e293b');

      let currentX = margin;
      tableHeaders.forEach((h, i) => {
        const w = colWidths[i];
        doc.fillColor('#ffffff')
          .font('Helvetica-Bold')
          .fontSize(8)
          .text(String(h).toUpperCase(), currentX + 4, startY + 6, {
            width: w - 8,
            align: (i === 0 || h.toString().toLowerCase().includes('score')) ? 'center' : 'left',
            lineBreak: false
          });
        currentX += w;
      });

      startY += headerHeight;

      // Draw Table Rows
      const rowHeight = 20;
      const maxY = pageHeight - margin - 35;

      tableRows.forEach((row, rowIndex) => {
        if (startY + rowHeight > maxY) {
          doc.addPage();
          startY = margin + 10;

          // Repeat Header on new page
          doc.rect(margin, startY, contentWidth, headerHeight).fill('#1e293b');
          let pageX = margin;
          tableHeaders.forEach((h, i) => {
            const w = colWidths[i];
            doc.fillColor('#ffffff')
              .font('Helvetica-Bold')
              .fontSize(8)
              .text(String(h).toUpperCase(), pageX + 4, startY + 6, {
                width: w - 8,
                align: (i === 0 || h.toString().toLowerCase().includes('score')) ? 'center' : 'left',
                lineBreak: false
              });
            pageX += w;
          });
          startY += headerHeight;
        }

        // Alternating row background
        const bg = rowIndex % 2 === 0 ? '#ffffff' : '#f8fafc';
        doc.rect(margin, startY, contentWidth, rowHeight).fill(bg).stroke('#e2e8f0');

        let cellX = margin;
        row.forEach((cellVal, cellIdx) => {
          const w = colWidths[cellIdx];
          let valStr = String(cellVal !== undefined && cellVal !== null ? cellVal : '');

          if (valStr === 'PRESENT') {
            doc.fillColor('#16a34a').font('Helvetica-Bold').fontSize(8);
          } else if (valStr === 'ABSENT') {
            doc.fillColor('#dc2626').font('Helvetica-Bold').fontSize(8);
          } else {
            doc.fillColor('#334155').font('Helvetica').fontSize(8);
          }

          const isCenter = (cellIdx === 0 || tableHeaders[cellIdx].toString().toLowerCase().includes('score') || valStr === 'PRESENT' || valStr === 'ABSENT');
          doc.text(valStr, cellX + 4, startY + 5, {
            width: w - 8,
            align: isCenter ? 'center' : 'left',
            lineBreak: false
          });
          cellX += w;
        });

        startY += rowHeight;
      });

      // Signature & Coordinator Signoff at end of table
      if (startY + 45 > maxY) {
        doc.addPage();
        startY = margin + 20;
      } else {
        startY += 20;
      }

      doc.fillColor('#1e293b').font('Helvetica-Bold').fontSize(8.5);
      doc.text(`Event Coordinator: ${coordinator || 'Unassigned'}`, margin + 10, startY);
      doc.text(`Signature: ___________________________`, contentWidth - 190, startY);
    }

    // Add Page Numbers in Footer for all pages
    const range = doc.bufferedPageRange();
    for (let i = range.start; i < range.start + range.count; i++) {
      doc.switchToPage(i);
      doc.fillColor('#64748b')
        .font('Helvetica')
        .fontSize(7.5)
        .text(
          `Tech Manthan 6.0 Report  |  Generated on: ${new Date().toLocaleString('en-IN')}  |  Page ${i + 1} of ${range.count}`,
          margin,
          pageHeight - margin + 10,
          { width: contentWidth, align: 'center' }
        );
    }

    doc.end();
  } catch (err) {
    console.error('PDF Generation Error:', err);
    if (!res.headersSent) {
      res.status(500).json({ error: 'Failed to generate PDF document', details: err.message });
    }
  }
};
